store/src/java/com/zimbra/cs/mailbox/DistributedMailboxCacheManager.java

```diff
@@ -0,0 +1,220 @@
+     */
+    private MailboxMap cache;
+
+    public DistributedMailboxCacheManager() {
+        DbPool.DbConnection conn = null;
+        synchronized (this) {
+            try {
+                conn = DbPool.getConnection();
+                mailboxIds = DbMailbox.listMailboxes(conn, null);
+                cache = createCache();
+            } catch (ServiceException e) {
+                ZimbraLog.mailbox.error("creating cache manager: ", e);
+                throw new CacheManagerException("creating cache manager: ", e.getCause());
+            } finally {
+                DbPool.quietClose(conn);
+            }
+        }
+    }
+
+    private synchronized Map<String, Integer> listMailboxes() {
+        DbPool.DbConnection conn = null;
+        try {
+            return DbMailbox.listMailboxes(conn, null);
+        } catch (ServiceException e) {
+            ZimbraLog.mailbox.error("fetching mailboxes: ", e);
+            throw new CacheManagerException("fetching mailboxes: ", e.getCause());
+        } finally {
+            DbPool.quietClose(conn);
+        }
+    }
+
+    @Override
+    public int getMailboxCount(){
+        Integer mailboxCount = null;
+        DbPool.DbConnection conn = null;
+        try {
+            try {
+                conn = DbPool.getConnection();
+                mailboxCount =  DbMailbox.getMailboxCount(conn);
+            } catch (ServiceException e) {
+                ZimbraLog.mailbox.error("getting mailbox count: ", e);
+                throw new CacheManagerException("getting mailbox count: ", e.getCause());
+            }
+        }finally {
+            DbPool.quietClose(conn);
+        }
+        return mailboxCount;
+    }
+
+    @Override
+    public Set<String> getAccountIds(){
+        return listMailboxes().keySet();
+    }
+
+    @Override
+    public Set<Map.Entry<String, Integer>> getMailboxIdsByAccountId() {
+        return listMailboxes().entrySet();
+    }
+
+    public Collection<Integer> getMailboxIds(){
+        return listMailboxes().values();
+    }
+
+
+    @Override
+    public Integer getMailboxKey(String accountId){
+        Integer mailboxKey = null;
+        DbPool.DbConnection conn = null;
+        try {
+            try {
+                conn = DbPool.getConnection();
+                mailboxKey =  DbMailbox.getMailboxKey(conn, accountId);
+            } catch (ServiceException e) {
+                ZimbraLog.mailbox.error("fetching mailbox key: ", e);
+                throw new CacheManagerException("fetching mailbox key: ", e.getCause());
+            }
+        }finally {
+            DbPool.quietClose(conn);
+        }
+        return mailboxKey;
+    }
+
+    @Override
+    public void removeMailboxId(String accountId) {
+        //noop
+    }
+
+    @Override
+    public Object retrieveFromCache(int mailboxId, boolean trackGC, MailboxManager mailboxManager) throws MailServiceException {
```

Comment:
`mailboxManager` does not appear to be used here

Reply:
Yeah is not being used, I left it for being able to extract this method into the `MailboxCacheManager` interface  . `mailboxManager`  is being used in `LocalMailboxCacheManager.retrieveFromCache`. Please let me know I need to do another refactoring here